web/src/compose_reply.js

```diff
@@ -129,10 +130,35 @@ export function reply_with_mention(opts) {
     compose_ui.insert_syntax_and_focus(mention);
 }
 
+export function selection_within_message_id(selection = window.getSelection()) {
+    // Returns the message_id if the selection is entirely within a message,
+    // otherwise returns undefined.
+    if (!selection.toString()) {
+        return undefined;
+    }
+    const {start_id, end_id} = copy_and_paste.analyze_selection(selection);
+    if (start_id === end_id) {
+        return start_id;
+    }
+    return undefined;
+}
+
 export function quote_and_reply(opts) {
+    let message_id;
+    let quote_content;
+    if (opts.message_id) {
+        message_id = opts.message_id;
+        quote_content = opts.quote_content;
+    } else if (selection_within_message_id()) {
+        // If this was triggered via hotkey, and the current selection is
+        // entirely within a message, we quote that selection.
+        message_id = selection_within_message_id();
+        quote_content = get_message_selection();
+    }
     assert(message_lists.current !== undefined);
-    const message_id = opts.message_id || message_lists.current.selected_id();
+    message_id ??= message_lists.current.selected_id();
     const message = message_lists.current.get(message_id);
+    quote_content ??= message.raw_content;
```

Comment:
I don't love the use of the `??=` operator here -- write a proper `if` block -- it feels hard to read. 

Also you may benefit from using more variable names in this block. For example, rather than calling `selection_within_message_id()` twice, just define a `const selection_message_id = ...`. So this could look something like the following, which very clearly lays out the 3 cases:

```
if (opts.message_id) {
     message_id = opts.message_id;
     quote_content = opts.quote_content;
} else {
    const selection_message_id = selection_within_message_id();
    if (selection_message_id !== undefined) {
         message_id = selection_message_id;
         quote_content = get_message_selection();
    } else {
         message = message_lists.current.selected_message();
         assert message !== undefined;
         quote_content = message.raw_content;
         message_id = message.id
    }

}


```

Reply:
@timabbott Did some clean up, and added some comments too. If it's readable enough now, I think this should be ready for CZO